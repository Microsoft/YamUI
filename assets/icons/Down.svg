<svg width="240" height="240" viewBox="0 0 240 240" xmlns="http://www.w3.org/2000/svg">
  <path d="M230.273 129.023L120 239.297 9.727 129.023l10.546-10.546 92.227 92.226V0h15v210.703l92.227-92.226z" fill-rule="evenodd"/>
</svg>
