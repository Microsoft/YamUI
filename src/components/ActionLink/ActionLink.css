.y-actionLink--wrapper {
  padding: 0.2rem 0 0.6rem;
}

.y-actionLink--icon {
  margin-top: 0.4rem;

  /* Remove when this is addressed: https://github.com/Microsoft/YamUI/issues/327 */
  position: static;
  top: 0;
}
