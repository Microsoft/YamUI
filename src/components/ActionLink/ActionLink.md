### Examples


With onClick (button implementation):

```js { "props": { "data-description": "with onClick", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-clickable\"}]" } }
const Icon = require('../Icon/icons/Download').default;

<div>
  <ActionLink text="Download" icon={Icon} ariaLabel="Download this file" onClick={action('ActionLink clicked')} />
</div>
```

With long text and onClick:

```js { "props": { "data-description": "with onClick and long text" } }
const Icon = require('../Icon/icons/Upload').default;
const text = 'Upload a new version of this file';

<div style={{ maxWidth: '200px' }}>
  <ActionLink text={text} icon={Icon} ariaLabel={text} onClick={action('ActionLink clicked')} />
</div>
```

With href (link implementation):

```js { "props": { "data-description": "with href" } }
const Icon = require('../Icon/icons/Group').default;

<div>
  <ActionLink text="Group Info" icon={Icon} ariaLabel="Go to group info page" href="https://www.yammer.com" newWindow={true} />
</div>
```

With long text and href:

```js { "props": { "data-description": "with href and long text" } }
const Icon = require('../Icon/icons/SharePointLogoInverse').default;
const text = 'SharePoint Document Library website';

<div style={{ maxWidth: '200px' }}>
  <ActionLink text={text} icon={Icon} ariaLabel={text} href="https://www.yammer.com" newWindow={true} />
</div>
```

Compact:

```js { "props": { "data-description": "with compact" } }
const Like = require('../Icon/icons/Like').default;

<div>
  <ActionLink text="Like" icon={Like} ariaLabel="Like this Post" compact={true} />
</div>
```
