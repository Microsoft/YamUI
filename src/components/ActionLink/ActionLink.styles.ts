/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { mergeStyleSets } from 'office-ui-fabric-react/lib/Styling';
import { memoizeFunction } from 'office-ui-fabric-react/lib/Utilities';

export const getClassNames = memoizeFunction(() => {
  return mergeStyleSets({
    wrapper: {
      padding: '0.2rem 0 0.6rem',
    },
  });
});
