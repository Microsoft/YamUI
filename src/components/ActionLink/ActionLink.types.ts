/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { BaseComponentProps } from '../../util/BaseComponent/props';
import { BaseIcon } from '../Icon';

export interface BaseActionLinkProps extends BaseComponentProps {
  /**
   * The icon component class to be rendered.
   */
  icon: typeof BaseIcon;

  /**
   * A well-curated string describing the action for screen readers.
   */
  ariaLabel?: string;

  /**
   * The visible text.
   */
  text: string;
}

export interface NavigationActionLinkProps extends BaseActionLinkProps {
  /**
   * A valid URL.
   */
  href: string;

  /**
   * Whether the link should open in a new window. It securely removes access to the opening window.
   */
  newWindow?: boolean;
}

export interface ClickableActionLinkProps extends BaseActionLinkProps {
  /**
   * A click handler.
   */
  onClick: ((e: React.MouseEvent<HTMLButtonElement | HTMLLinkElement>) => void);
}

export type ActionLinkProps = NavigationActionLinkProps | ClickableActionLinkProps;
