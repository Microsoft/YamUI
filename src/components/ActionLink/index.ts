/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export { default } from './ActionLink';
export * from './ActionLink.types';
