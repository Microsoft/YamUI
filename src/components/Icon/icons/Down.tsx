/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
/* tslint:disable:max-line-length */
import * as React from 'react';

const Down = function Down(props: any) {
  return (
    <svg xmlns="http://www.w3.org/2000/svg" width={320} height={320} viewBox="0 0 320 320" {...props}>
      <path d="M160 195.774L64.072 99.846A19.9 19.9 0 0 0 49.96 94a19.9 19.9 0 0 0-14.114 5.846c-7.795 7.795-7.795 20.432 0 28.226L144.803 237.03c.33.384.68.76 1.043 1.124A19.895 19.895 0 0 0 160 244a19.897 19.897 0 0 0 14.154-5.846c.365-.365.712-.74 1.043-1.125L284.154 128.07c7.795-7.794 7.795-20.43 0-28.226A19.906 19.906 0 0 0 270.04 94a19.9 19.9 0 0 0-14.112 5.846L160 195.774z" />
    </svg>
  );
};

export default Down;
