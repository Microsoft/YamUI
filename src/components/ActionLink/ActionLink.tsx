/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import { join } from '../../util/classNames';
import Block, { TextSize } from '../Block';
import Clickable from '../Clickable';
import { FixedGridRow, FixedGridColumn, GutterSize } from '../FixedGrid';
import { BaseIcon, IconSize } from '../Icon';
import NavigationLink from '../NavigationLink';
import './ActionLink.css';

export interface BaseActionLinkProps extends BaseComponentProps {
  /**
   * The icon component class to be rendered.
   */
  icon: typeof BaseIcon;

  /**
   * A well-curated string describing the action for screen readers.
   */
  ariaLabel: string;

  /**
   * The visible text.
   */
  text: string;
}

export interface NavigationActionLinkProps extends BaseActionLinkProps {
  /**
   * A valid URL.
   */
  href: string;

  /**
   * Whether the link should open in a new window. It securely removes access to the opening window.
   */
  newWindow?: boolean;
}

export interface ClickableActionLinkProps extends BaseActionLinkProps {
  /**
   * A click handler.
   */
  onClick: ((e: React.MouseEvent<HTMLButtonElement | HTMLLinkElement>) => void);
}

export type ActionLinkProps = NavigationActionLinkProps | ClickableActionLinkProps;

/**
 * An actionable element with icon and text which will either render a button with an onClick callback
 * or a navigation link.
 */
export default class ActionLink extends React.Component<ActionLinkProps> {
  public render() {
    const { ariaLabel, className, icon: Icon, text } = this.props;
    const classNames = join(['y-actionLink', className]);

    const content = (
      <FixedGridRow gutterSize={GutterSize.SMALL} className="y-actionLink--wrapper">
        <FixedGridColumn fixed={true}>
          <Icon size={IconSize.MEDIUM} block={true} className="y-actionLink--icon" />
        </FixedGridColumn>
        <FixedGridColumn>
          <Block textSize={TextSize.MEDIUM_SUB}>{text}</Block>
        </FixedGridColumn>
      </FixedGridRow>
    );

    if ((this.props as NavigationActionLinkProps).href) {
      return (
        <NavigationLink
          href={(this.props as NavigationActionLinkProps).href}
          newWindow={(this.props as NavigationActionLinkProps).newWindow}
          ariaLabel={ariaLabel}
          className={classNames}
        >
          {content}
        </NavigationLink>
      );
    }
    return (
      <Clickable
        onClick={(this.props as ClickableActionLinkProps).onClick}
        ariaLabel={ariaLabel}
        className={classNames}
        block={true}
      >
        {content}
      </Clickable>
    );
  }
}
