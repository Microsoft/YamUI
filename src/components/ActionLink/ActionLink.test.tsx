/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import ActionLink, { ActionLinkProps } from './index';
import Down from '../Icon/icons/Down';

describe('<ActionLink />', () => {
  let component: ShallowWrapper<ActionLinkProps>;

  describe('with href', () => {
    beforeEach(() => {
      component = shallow(<ActionLink text="Download" icon={Down} href="https://yammer.com" newWindow={true} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with href and ariaLabel', () => {
    beforeEach(() => {
      component = shallow(
        <ActionLink text="Download" icon={Down} ariaLabel="Download file" href="https://yammer.com" newWindow={true} />,
      );
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with onClick', () => {
    let onClick: jest.Mock;

    beforeEach(() => {
      onClick = jest.fn();
      component = shallow(<ActionLink text="Download" icon={Down} onClick={onClick} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    describe('when clicked', () => {
      beforeEach(() => {
        component.simulate('click');
      });

      it('calls props.onClick', () => {
        expect(onClick).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe('with onClick and ariaLabel', () => {
    let onClick: jest.Mock;

    beforeEach(() => {
      onClick = jest.fn();
      component = shallow(<ActionLink text="Download" ariaLabel="Download file" icon={Down} onClick={onClick} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(
        <ActionLink
          className="TEST_CLASSNAME"
          text="Download"
          icon={Down}
          ariaLabel="Download the file"
          href="https://yammer.com"
        />,
      );
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with compact set to true', () => {
    beforeEach(() => {
      component = shallow(
        <ActionLink
          className="TEST_CLASSNAME"
          text="Download"
          icon={Down}
          ariaLabel="Download the file"
          href="https://yammer.com"
          compact={true}
        />,
      );
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
