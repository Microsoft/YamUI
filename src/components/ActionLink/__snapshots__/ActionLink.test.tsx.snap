// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionLink /> with additional className matches its snapshot 1`] = `
<CustomizedNavigationLink
  ariaLabel="Download the file"
  block={true}
  className="y-actionLink TEST_CLASSNAME"
  href="https://yammer.com"
>
  <FixedGridRow
    gutterSize="small"
  >
    <FixedGridColumn
      fixed={true}
    >
      <CustomizedBlock
        push={2}
      >
        <Down
          block={true}
          size="16"
        />
      </CustomizedBlock>
    </FixedGridColumn>
    <FixedGridColumn>
      <CustomizedBlock
        textSize="mediumSub"
      >
        Download
      </CustomizedBlock>
    </FixedGridColumn>
  </FixedGridRow>
</CustomizedNavigationLink>
`;

exports[`<ActionLink /> with compact set to true matches its snapshot 1`] = `
<CustomizedNavigationLink
  ariaLabel="Download the file"
  block={true}
  className="y-actionLink TEST_CLASSNAME"
  href="https://yammer.com"
>
  <FixedGridRow
    gutterSize="xSmall"
  >
    <FixedGridColumn
      fixed={true}
    >
      <CustomizedBlock
        push={2}
      >
        <Down
          block={true}
          size="14"
        />
      </CustomizedBlock>
    </FixedGridColumn>
    <FixedGridColumn>
      <CustomizedBlock
        textSize="mediumSub"
      >
        <Strong>
          Download
        </Strong>
      </CustomizedBlock>
    </FixedGridColumn>
  </FixedGridRow>
</CustomizedNavigationLink>
`;

exports[`<ActionLink /> with href and ariaLabel matches its snapshot 1`] = `
<CustomizedNavigationLink
  ariaLabel="Download file"
  block={true}
  className="y-actionLink"
  href="https://yammer.com"
  newWindow={true}
>
  <FixedGridRow
    gutterSize="small"
  >
    <FixedGridColumn
      fixed={true}
    >
      <CustomizedBlock
        push={2}
      >
        <Down
          block={true}
          size="16"
        />
      </CustomizedBlock>
    </FixedGridColumn>
    <FixedGridColumn>
      <CustomizedBlock
        textSize="mediumSub"
      >
        Download
      </CustomizedBlock>
    </FixedGridColumn>
  </FixedGridRow>
</CustomizedNavigationLink>
`;

exports[`<ActionLink /> with href matches its snapshot 1`] = `
<CustomizedNavigationLink
  block={true}
  className="y-actionLink"
  href="https://yammer.com"
  newWindow={true}
>
  <FixedGridRow
    gutterSize="small"
  >
    <FixedGridColumn
      fixed={true}
    >
      <CustomizedBlock
        push={2}
      >
        <Down
          block={true}
          size="16"
        />
      </CustomizedBlock>
    </FixedGridColumn>
    <FixedGridColumn>
      <CustomizedBlock
        textSize="mediumSub"
      >
        Download
      </CustomizedBlock>
    </FixedGridColumn>
  </FixedGridRow>
</CustomizedNavigationLink>
`;

exports[`<ActionLink /> with onClick and ariaLabel matches its snapshot 1`] = `
<CustomizedClickable
  ariaLabel="Download file"
  block={true}
  className="y-actionLink"
  onClick={[MockFunction]}
>
  <FixedGridRow
    gutterSize="small"
  >
    <FixedGridColumn
      fixed={true}
    >
      <CustomizedBlock
        push={2}
      >
        <Down
          block={true}
          size="16"
        />
      </CustomizedBlock>
    </FixedGridColumn>
    <FixedGridColumn>
      <CustomizedBlock
        textSize="mediumSub"
      >
        Download
      </CustomizedBlock>
    </FixedGridColumn>
  </FixedGridRow>
</CustomizedClickable>
`;

exports[`<ActionLink /> with onClick matches its snapshot 1`] = `
<CustomizedClickable
  block={true}
  className="y-actionLink"
  onClick={[MockFunction]}
>
  <FixedGridRow
    gutterSize="small"
  >
    <FixedGridColumn
      fixed={true}
    >
      <CustomizedBlock
        push={2}
      >
        <Down
          block={true}
          size="16"
        />
      </CustomizedBlock>
    </FixedGridColumn>
    <FixedGridColumn>
      <CustomizedBlock
        textSize="mediumSub"
      >
        Download
      </CustomizedBlock>
    </FixedGridColumn>
  </FixedGridRow>
</CustomizedClickable>
`;
